### PR TITLE
First attempt at a docker development environment

### PR DESCRIPTION
**Pull request by ghost**
This is the promised docker development environment

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -11,3 +11,9 @@ SOCIALHOME_DOMAIN=127.0.0.1:8000
 SOCIALHOME_HTTPS=False
 
 MOCHA_TESTS=True
+
+# Set these variables for Docker Postfix (be sure the DATABASE_URL matches)
```
Comment (jaywink) [inline .env.example]: `Postfix` -> `PostgreSQL`.

I'm also wondering why do we need these here at all? Why would anyone want to change these for a development database? IMHO, they should just be set in `docker-compose.yml`. If someone *wants* to change these, they can change that file - it's a local copy anyway not part of the repo. If possible, it would be nice to not have extra optional stuff in `.env.example` - it already has a few unnecessary things that should be cleaned out.

File: docker/dev/Dockerfile.django [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# We use the official Python images. This may need to get locked in, or back to Ubuntu at some point
```
Comment (jaywink) [inline docker/dev/Dockerfile.django]: Maybe this commetn can be removed if the challenge has already been solved :)

File: docker/dev/Dockerfile.django [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+FROM python:3
+
+RUN apt-get update
+RUN apt-get install -y lib32z1-dev lib32z1 libpq-dev
```
Comment (jaywink) [inline docker/dev/Dockerfile.django]: I'm kind of surprised `libxml2-dev libxslt-dev` are not needed. Those are direct dependencies for `lxml` to be built successfully. Maybe the Python3 image has them already?

File: docker/dev/Dockerfile.npm [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# We use the official Python images. This may need to get locked in, or back to Ubuntu at some point
```
Comment (jaywink) [inline docker/dev/Dockerfile.npm]: Redundant comment

File: docker/dev/Dockerfile.npm [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+FROM node
+
+RUN mkdir /code
+
+WORKDIR /code
+COPY . .
+
+RUN npm install
+RUN node_modules/.bin/bower --allow-root install
+RUN node_modules/.bin/grunt build
```
Comment (jaywink) [inline docker/dev/Dockerfile.npm]: `grunt dev` when doing development - this ensures the scripts are not minified, helping debug problems.

~~Also, `npm run dev` is missing. This builds the webpack bundle. Failing to do so will crash the UI when trying to access a Vue view.~~

File: docker/dev/docker-entrypoint.sh.django [hunk 6]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
+#
+# Socialhome Docker Development entrpypoint
```
Comment (jaywink) [inline docker/dev/docker-entrypoint.sh.django]: Typo in "entrypoint"

File: docs/development.rst [hunk 7]
```diff
@@ -63,11 +63,11 @@ To watch files and build bundles automatically, use this.
 Configure
 .........
 
-Configuration is done via environment variables. For the meaning of them, look them up under files in ``config/settings``. Values in the file ``.env`` will be used automatically.
+Configuration is done via environment variables. For the meaning of them, look them up under files in ``config/settings``. Values in ``env.local`` will be used automatically.
```
Comment (jaywink) [inline docs/development.rst]: This is reverting my previous commit. I think you somehow managed to copy the old file with your changes, thus reverting some other stuff I committed since. This is why it is important to develop in branches, not in the master branch.

File: docs/development.rst [hunk 8]
```diff
@@ -63,11 +63,11 @@ To watch files and build bundles automatically, use this.
...
 
 ::
 
-    cp .env.example .env
+    cp env.example env.local
```
Comment (jaywink) [inline docs/development.rst]: Reverting master changes here too.

File: docs/development.rst [hunk 9]
```diff
@@ -148,18 +148,6 @@ Execute the following to run the new frontend JavaScript tests.
 
     npm run test
 
-API routes
```
Comment (jaywink) [inline docs/development.rst:151]: Reverting master changes here too.

File: docs/development.rst [hunk 10]
```diff
@@ -203,6 +191,42 @@ Then execute the following and copy the markdown version for pasting to GitHub r
 
 After the release commit has been pushed and a release has been tagged, set a development version in the same above files. This is basically the next minor release postfixed by ``-dev``.
 
+Developing with Docker
+------------------------
+
+If you choose, you may develop Socialhome using Docker, rather than installing Postfix and Redis manually on your computer.
```
Comment (jaywink) [inline docs/development.rst]: `Postfix` -> `PostgreSQL`.

Maybe someday we *will* have a postfix dependency to allow Socialhome to act as a mail server ;)

File: docs/development.rst [hunk 11]
```diff
@@ -203,6 +191,42 @@ Then execute the following and copy the markdown version for pasting to GitHub r
...
+And then the steps you would normally do, but throught he django image, ala:
+
+``docker-compose run django manage migrate``
+and
+``docker-compose run django manage createsuperuser``
+
+And then just
+
+``docker-compose up``
+
```
Comment (jaywink) [inline docs/development.rst]: I wonder if it would make sense to also mention where the UI is exposed to, ie what to type in the browser to reach the running instance? Might help not so experienced developers. It's not apparent to me at least where I should look.

File: docs/development.rst [hunk 12]
```diff
@@ -203,6 +191,42 @@ Then execute the following and copy the markdown version for pasting to GitHub r
...
+..................
+
+This guide assumes you are running Docker on a GNU/Linux based system such as Ubuntu, Debian or Fedora Linux. It may be possible to run this on other platforms where Docker is supported, but those are untested.
+
+The docker development installation was tested on Docker version 17.09 and docker-compose 1.16.1.
+
+Steps
+......
+
+The first step is to copy the ``docker/dev/docker-compose.yml`` file to the root of the project. eg
```
Comment (jaywink) [inline docs/development.rst]: The file is missing from this commit, probably need to tune the `.gitignore` to exlude it from that path, but include it when in root path.

**Review by jaywink — CHANGES_REQUESTED** [review]
Great work, thanks for this! 🎉  Some comments to move forward.

**Comment by jaywink** [comment]
Also, as a generic comment. Always use a branch in your own fork and do PR's from that branch to upstream master, This makes mistakes like overwriting upstream changes less likely and also allows to work on multiple things at once 👍

**Comment by ghost** [comment]
> Also, as a generic comment. Always use a branch in your own fork and do PR's from that branch to upstream master

Was using a local branch locally but good point.

Let me address the others then see what I need to do to ammend the PR

File: .env.example (hunk 1 above)
Comment (ghost) [reply .env.example]: This is a good question. Here's my reasoning:

The .env file is also used by docker-compose for its environment variables. So this means we only have one source of truth. But it's easily moved to the docker-compose file.

File: docker/dev/Dockerfile.django [hunk 13]
```diff
@@ -1,12 +1,6 @@
-# We use the official Python images. This may need to get locked in, or back to Ubuntu at some point
 FROM python:3
 
-RUN apt-get update
```
Comment (jaywink) [inline docker/dev/Dockerfile.django]: Hmm, why did you remove the package installs?

Comment (ghost) [reply docker/dev/Dockerfile.django]: The reason for the package installs was just for XSLT and related.

Otherwise, the Docker image software should be maintained by the upsteam Docker maintainers. There are three reasons why this is the case:

1. This is the Docker data model

What I mean is that Docker images will only fetch data that's necessary to overlay on top of each other, recursively. It's more space efficient to do this

2. We're using the official Python docker images

Just as we would generally rely on upstream package maintainers in Linux or BSD to maintain upstream code, we should be leaning on the Python docker maintainers to maintain upstream docker. The only reason we pulled in additional packages was for XSLT, and as you mentioned, that's no longer necessary.

3. If the official Python image changes somehow, we're less likely to break.

It appears that Python's docker image is using Debian, but if it should change somehow, our docker images will still build properly if we aren't relying on anything other than Python itself.

That's why I removed it.

**Comment by ghost** [comment]
I believe the last remaining objection was the removal of the apt-get lines from the Django image, which I've explained.

The only other changes I've made since the origial PR were to address the issues @jaywink  mentioned in his comments as well as add some additional verbosity to the Javascript watching image.

File: docker/dev/Dockerfile.django (hunk 13 above)
Comment (jaywink) [reply docker/dev/Dockerfile.django]: Actually it was a question:

> I'm kind of surprised libxml2-dev libxslt-dev are not needed. Those are direct dependencies for lxml to be built successfully. Maybe the Python3 image has them already?

https://github.com/jaywink/socialhome/pull/334#discussion_r145684629

At least Ubuntu requires a few extra libraries to build `lxml` (see http://lxml.de/installation.html#requirements and https://socialhome.readthedocs.io/en/latest/install_guides.html#install-system-packages (federation ie lxml dependencies)). So if the `python:3` image does work without these extra installs, it must have them already? Otherwise I can't imagine that the docker images work.

Does a clean build of this docker image without these extra installs actually work? And tests also pass?

File: docs/development.rst (hunk 12 above)
Comment (jaywink) [reply docs/development.rst]: `docker-compose.yml` is still missing from the PR.

File: .env.example [hunk 14]
```diff
@@ -11,3 +11,9 @@ SOCIALHOME_DOMAIN=127.0.0.1:8000
 SOCIALHOME_HTTPS=False
 
 MOCHA_TESTS=True
+
+# Set these variables for Docker Postgres (be sure the DATABASE_URL matches)
+POSTGRES_USER=socialhome
```
Comment (jaywink) [inline .env.example]: As discussed, these variables should be imho defined in `docker-compose.yml` (which is missing btw). Otherwise we have database settings twice, and really there is no reason for developers normally to change these.

**Comment by jaywink** [comment]
There are still some master commit things being reverted in this PR - please check the comments carefully, thanks.

**Comment by jaywink** [comment]
@Vir-Cotto could you check the remaining comments I've made, for example using this view: https://github.com/jaywink/socialhome/pull/334/files .. all of those are still relevant. Thanks!

File: docs/development.rst (hunk 8 above)
Comment (ghost) [reply docs/development.rst]: :heavy_check_mark:

File: docs/development.rst (hunk 9 above)
Comment (ghost) [reply docs/development.rst:151]: :heavy_check_mark:  Sorry about that.

File: docs/development.rst (hunk 10 above)
Comment (ghost) [reply docs/development.rst]: Yes, this is actually a thing to think about in production (since notifications are via mail) but it's fixed in my current version. :heavy_check_mark:

File: docs/development.rst (hunk 12 above)
Comment (ghost) [reply docs/development.rst]: It's on the .gitignore and in the docker/dev directory. :heavy_check_mark:

File: .env.example (hunk 14 above)
Comment (ghost) [reply .env.example]: :heavy_check_mark:

File: docs/development.rst (hunk 11 above)
Comment (ghost) [inline docs/development.rst:241]: Done. :heavy_check_mark:

**Comment by ghost** [comment]
All the comments should now be addressed.

**Comment by jaywink** [comment]
I think you maybe forgot to push your changes? I see you've checked my comments but nothing has changed regarding the comments, sorry.

File: docs/development.rst [hunk 15]
```diff
@@ -203,6 +191,47 @@ Then execute the following and copy the markdown version for pasting to GitHub r
...
+``docker-compose run django manage createsuperuser``
+
+And then just
+
+``docker-compose up``
+
+Defaults
+..........
+
+The defaults are that that the Docker image will be running on port 8000 and then exposed to the host OS on the same port (ie you can browse to http;//localhost:8000 to see the Django instance running). Redis and Postfix will be running but not exposed to the host OS by default. These can be changed on the ``docker-compose.yml`` file.
```
Comment (jaywink) [inline docs/development.rst]: `Postfix` -> `PostgreSQL`

**Comment by ghost** [comment]
Fixed the Postfix mistypes. I have email on the brain.

**Comment by ghost** [comment]
The documentation is now fixed. I'm not sure what else needs addressing. I would recommend squashing the last few commits.

**Review by jaywink — APPROVED** [review]

**Comment by jaywink** [comment]
Thanks! 🎉
